packages/flutter/lib/src/painting/borders.dart

```diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
   ///  * [Path.contains], which can tell if an [Offset] is within a [Path].
   Path getInnerPath(Rect rect, { TextDirection? textDirection });
 
+  /// Paint a canvas with appropriate shape.
+  ///
+  /// This works as an optimization for [getOuterPath] + [Canvas.drawPath],
+  /// which can be slow. Drawing can be faster by providing a [paintInterior]
+  /// with a specific draw method (instead of a generic [Canvas.drawPath]).
+  /// Examples of specific methods are [Canvas.drawRect] and [Canvas.drawCircle].
+  ///
+  /// A standard paintBox implementation is `canvas.draw` followed by
+  /// what is inside [getOuterPath].
+  ///
+  /// As an example, see how [Path.addRect] became [Canvas.drawRect]:
+  ///
+  /// ```dart
+  /// @override
+  /// Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
+  ///  return Path()
+  ///    ..addRect(rect);
+  /// }
+  ///
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) {
+  ///  canvas.drawRect(rect, paint);
+  /// }
+  /// ```
+  ///
+  /// When a shape can only be drawn using path, [preferPaintInterior] is recommended
+  /// to be false, so [ShapeBorder] caches the path.
+  ///
+  /// This function provides a default empty implementation as it is only executed
+  /// when [preferPaintInterior] is true.
```

Comment:
I had a go at rewriting this to use the usual Flutter voice for documentation:

```suggestion
  /// Paint a canvas with the appropriate shape.
  ///
  /// On [ShapeBorder] subclasses whose [preferPaintInterior] method returns true,
  /// this should be faster than using [Canvas.drawPath] with the path provided by
  /// [getOuterPath]. (If [preferPaintInterior] returns false, then this method asserts
  /// in debug mode and does nothing in release mode.)
  ///
  /// Subclasses are expected to implement this method when the [Canvas] API has
  /// a dedicated method to draw the relevant shape. For example, [CircleBorder] uses
  /// this to call [Canvas.drawCircle], and [RoundedRectangleBorder] uses this to
  /// call [Canvas.drawRRect].
  ///
  /// Subclasses that implement this must ensure that calling [paintInterior] is semantically
  /// equivalent to (i.e. renders the same pixels as) calling [Canvas.drawPath] with the
  /// same [Paint] and the [Path] returned from [getOuterPath]), and must also override
  /// [preferPaintInterior] to return true.
  ///
  /// For example, a shape that draws a rectangle might implement [getOutputPath],
  /// [paintInterior], and [preferPaintInterior] as follows:
  ///
  /// ```dart
  /// class RectangleBorder extends ShapeBorder {
  ///   ...
  ///
  ///   @override
  ///   Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
  ///    return Path()
  ///      ..addRect(rect);
  ///   }
  ///
  ///   @override
  ///   void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) {
  ///    canvas.drawRect(rect, paint);
  ///   }
  ///
  ///   @override
￼  ///   bool get preferPaintInterior => true;
  /// }
  /// ```
  ///
  /// When a shape can only be drawn using path, [preferPaintInterior] should return false.
  /// In that case, classes such as [ShapeDecoration] will cache the path from [getOuterPath]
  /// and call [Canvas.drawPath] directly.
```

Reply:
(i didn't check that the wrapping is correct)